src/Features/Core/Portable/FileHeaders/AbstractFileHeaderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.IO;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FileHeaders
+{
+    internal abstract class AbstractFileHeaderDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        protected AbstractFileHeaderDiagnosticAnalyzer(string language)
+            : base(
+                IDEDiagnosticIds.FileHeaderMismatch,
+                CodeStyleOptions.FileHeaderTemplate,
+                language,
+                new LocalizableResourceString(nameof(FeaturesResources.The_file_header_is_missing_or_not_located_at_the_top_of_the_file), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.A_source_file_is_missing_a_required_header), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            RoslynDebug.AssertNotNull(DescriptorId);
+
+            var invalidHeaderTitle = new LocalizableResourceString(nameof(FeaturesResources.The_file_header_does_not_match_the_required_text), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            var invalidHeaderMessage = new LocalizableResourceString(nameof(FeaturesResources.A_source_file_contains_a_header_that_does_not_match_the_required_text), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            InvalidHeaderDescriptor = CreateDescriptorWithId(DescriptorId, invalidHeaderTitle, invalidHeaderMessage);
+        }
+
+        internal DiagnosticDescriptor MissingHeaderDescriptor => Descriptor;
+
+        internal DiagnosticDescriptor InvalidHeaderDescriptor { get; }
```

Comment:
do these need to be internal?  will protected not do?

Reply:
Analyzer descriptors are often needed by tests.